File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include <queue>
using namespace std;

queue<pair<int,int>>q;
int arr[101][101];
int dx[4]={0,0,1,-1};
int dy[4]={1,-1,0,0};
int answer = -1;
int n=0,m=0;

void bfs(){
    int visited[101][101] ={0};
    visited[0][0]=1;
    while(!q.empty()){
        int kx = q.front().first;
        int ky = q.front().second;
        q.pop();
        for(int i=0;i<4;i++){
            int nx = kx+dx[i];
            int ny = ky+dy[i];
            if(nx<0 || ny<0 ||nx>=n || ny>=m) continue;
            if(arr[nx][ny]==1 || visited[nx][ny]) continue;
            visited[nx][ny] = visited[kx][ky]+1;
            if(nx==n-1 && ny==m-1){
                answer = visited[nx][ny];
                return;
            }
            q.push({nx,ny});
        }
    }
}

int solution(vector<vector<int> > maps)
{
    n = maps.size();
    m = maps[0].size();
    for(int i=0;i<n;i++){
        for(int j=0;j<m;j++){
            if(maps[i][j]==0) arr[i][j]=1;
        }
    }
    
    q.push({0,0});
    bfs();
    
    return answer;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    private static class Des{
        int x;
        int y;
        int cnt;
        public Des(int x,int y, int cnt){
            this.x = x;
            this.y = y;
            this.cnt = cnt;
        }
    }
    
    private static int[] dx = {1,-1,0,0};
    private static int[] dy = {0,0,1,-1};
    private static boolean[][] visited;
    private static int answer = -1;
    
    
    private static void bfs(int n, int m, int[][] maps){
        Queue<Des> queue = new ArrayDeque<>();
        queue.add(new Des(0,0,1));
        visited = new boolean[n+1][m+1];
        visited[0][0]=true;
        
        while(!queue.isEmpty()){
            Des now = queue.poll();
            for(int i=0;i<4;i++){
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                int nc = now.cnt + 1;
                if(nx < 0 || ny < 0 || nx>=n || ny>=m) continue;
                if(visited[nx][ny]==true || maps[nx][ny]==0) continue;
                visited[nx][ny] = true;
                if(nx ==n-1 && ny ==m-1){
                    answer = nc;
                    return;
                }
                queue.add(new Des(nx,ny,nc));
                
            }
        }
        
    }
    
    public int solution(int[][] maps) {
        bfs(maps.length, maps[0].length, maps);
        
        return answer;
    }
}